odoo/models.py

```diff
@@ -1600,20 +1600,35 @@ def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get
     @api.model
     def _add_missing_default_values(self, values):
         # avoid overriding inherited values when parent is set
-        avoid_models = {
-            parent_model
-            for parent_model, parent_field in self._inherits.items()
-            if parent_field in values
-        }
+
+        def _get_avoid_models(model):
+            avoid_models = {
+                parent_model
+                for parent_model, parent_field in model._inherits.items()
+                if parent_field in values
+            }
+            avoid_models_inherited = []
+            for inherited_avoid_models in (_get_avoid_models(
+                self.env[parent_model])
+                for parent_model in model._inherits.keys()):
+                avoid_models_inherited.append(inherited_avoid_models)
+
+            return set().union(avoid_models, *avoid_models_inherited)
+
+        avoid_models = _get_avoid_models(self)
 
         # compute missing fields
-        missing_defaults = {
-            name
-            for name, field in self._fields.items()
-            if name not in values
-            if not (self._log_access and name in MAGIC_COLUMNS)
-            if not (field.inherited and field.related_field.model_name in avoid_models)
-        }
+        missing_defaults = set()
+        for name, field in self._fields.items():
+            if name not in values and \
+                    not (self._log_access and name in MAGIC_COLUMNS):
+                if field.inherited:
+                    aux_field = field
+                    while aux_field.inherited:
+                        aux_field = aux_field.related_field
+                    if aux_field.model_name in avoid_models:
+                        continue
+                missing_defaults.add(name)
```

Comment:
Your solution looks more complex than necessary, and I believe that `avoid_models` is not the right tool to discard inherited fields.  I suggest to build a blacklist of field names, and use it to filter out inherited fields that don't need a default:
```python
        # blacklist default values for fields inherited from a parent record
        blacklist = set()
        todo = [self]
        while todo:
            model = todo.pop()
            for parent_mname, parent_fname in model._inherits.items():
                parent_model = self.env[parent_mname]
                if parent_fname in values:
                    blacklist.update(parent_model._fields)
                else:
                    blacklist.difference_update(parent_model._fields)
                    todo.append(parent_model)

        # compute missing fields
        missing_defaults = [
            name
            for name, field in self._fields.items()
            if name not in values and not (field.inherited and name in blacklist)
        ]
```